analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs

```diff
@@ -0,0 +1,116 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseUnixEpochBase<TSyntaxKind, TLiteralExpression, TMemberAccessExpression> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TLiteralExpression : SyntaxNode
+    where TMemberAccessExpression : SyntaxNode
+{
+    private const string DiagnosticId = "S6588";
+
+    protected override string MessageFormat => "Use \"{0}.UnixEpoch\" instead of creating {0} instances that point to the unix epoch time";
+
+    protected abstract bool IsDateTimeKindUtc(TMemberAccessExpression memberAccess);
+    protected abstract bool IsGregorianCalendar(SyntaxNode node);
+    protected abstract bool IsZeroTimeOffset(SyntaxNode node);
+
+    protected UseUnixEpochBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(start =>
+        {
+            if (!CompilationTargetsValidNetVersion(start.Compilation))
+            {
+                return;
+            }
+
+            context.RegisterNodeAction(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var argumentCount = Language.Syntax.ArgumentExpressions(c.Node).Count();
+                    if (argumentCount < 3)
+                    {
+                        return;
+                    }
+
+                    var type = c.SemanticModel.GetTypeInfo(c.Node).Type;
+                    if (type.DerivesFrom(KnownType.System_DateTime) && IsEpochCtor(c.Node, c.SemanticModel))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation(), "DateTime"));
+                    }
+                    else if (type.DerivesFrom(KnownType.System_DateTimeOffset) && IsEpochCtor(c.Node, c.SemanticModel))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation(), "DateTimeOffset"));
+                    }
+                },
+                Language.SyntaxKind.ObjectCreationExpressions);
+        });
+
+    protected static bool IsValueEqualsTo(TLiteralExpression literal, int value) =>
+        literal.ChildTokens().First().ValueText == value.ToString();
+
+    private bool IsEpochCtor(SyntaxNode node, SemanticModel model)
+    {
+        var methodSymbol = (IMethodSymbol)model.GetSymbolInfo(node).Symbol;
```

Comment:
Do we need a symbolic lookup here in every case? This can be done completely with the syntax tree (even though it will have a more complex code). Or you can do a quick syntax check: does the method have at least one argument with the value of `1970` and at least two with the value of `1`? Only check the semantic mode if the answer to that is yes.

Reply:
If you want to do that, you should add a test case with `new DateTime(1970, 01, 01)`, because otherwise you will have some false negatives.

Reply:
The rule is raising correctly for `new DateTime(1970, 01, 01)`.
The syntax check on the 1/1/1970 arguments will already filter out most of the constructors, so I would keep it as is for the rest to be 100% sure the type is indeed _System.DateTime_. The overall perf cost will be very small and we will raise with more confidence.

Reply:
@cristian-ambrosini-sonarsource I don't see a test case for it though?

Reply:
You can find it in the `EdgeCases` method.